core/cog-view.c

```diff
@@ -378,6 +397,33 @@ cog_view_set_use_key_bindings(CogView *self, gboolean enable)
     g_object_notify_by_pspec(G_OBJECT(self), s_properties[PROP_USE_KEY_BINDINGS]);
 }
 
+/**
+ * cog_view_set_viewport: (set-property viewport)
+ * @self: A view.
+ * @viewport: The viewport where the View belongs.
+ *
+ * Since: 0.20
+ */
+void
+cog_view_set_viewport(CogView *self, CogViewport *viewport)
+{
+    g_return_if_fail(COG_IS_VIEW(self));
+
+    CogViewPrivate *priv = cog_view_get_instance_private(self);
+
+    if (priv->viewport != viewport) {
+        // If the new value is different from the old one, unreference the old value and set the new one.
+        if (viewport)
+            g_object_ref(viewport);
+        if (priv->viewport)
```

Comment:
This introduces a circular reference: the view references the viewport, and the viewport its views. Reference cycles are tricky, you would need to add add a custom `GObject.dispose` vfunc implementation that breaks the cycle and the object needs to allow public API function calls while in half-dead state while `priv->viewport` is null.

On the other hand, we have to allow `priv->viewport` to be null anyway, because the API has one function to create a view, and a different one to add (or remove) to a viewport. Taking that into account, I think it would be better to make `priv->viewport` a weak reference: this will make the `priv->viewport` pointer set to `NULL` when the viewport gets destroyed, and we do not need to handle cyclic references.

(As a bonus, using the weak reference approach we would allow programs which use the API to check if the viewport is set to `NULL` in a signal handler, and move the orphaned view to some other viewport. But this is not be the main reason to use a weak reference, I'd say.)

Reply:
I think this particular code are safe of this situation as it is but it is true that then I could have the problem in the for-each (https://github.com/Igalia/cog/pull/666#discussion_r1400789045). Totally agree.